hashes/src/sha512_256.rs

```diff
@@ -0,0 +1,262 @@
+// Bitcoin Hashes Library
+// Written in 2018 by
+//   Andrew Poelstra <apoelstra@wpsoftware.net>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+// This module is largely copied from the rust-crypto ripemd.rs file;
+// while rust-crypto is licensed under Apache, that file specifically
+// was written entirely by Andrew Poelstra, who is re-licensing its
+// contents here as CC0.
+
+//! SHA512_256 implementation.
+//!
+
+use core::{cmp, str};
+use core::ops::Index;
+use core::slice::SliceIndex;
+
+use crate::{Error, sha512, HashEngine as _, hex};
+
+const BLOCK_SIZE: usize = 128;
+
+/// Engine to compute SHA512/256 hash function.
+#[derive(Clone)]
+pub struct HashEngine (sha512::HashEngine);
+
+impl Default for HashEngine {
+    fn default() -> Self {
+        HashEngine(sha512::HashEngine {
+            h: [
+                0x22312194fc2bf72c, 0x9f555fa3c84c64c2, 0x2393b86b6f53b151, 0x963877195940eabd,
+                0x96283ee2a88effe3, 0xbe5e1e2553863992, 0x2b0199fc2c85b8aa, 0x0eb72ddc81c52ca2,
+            ],
+            length: 0,
+            buffer: [0; BLOCK_SIZE],
+        })
+    }
+}
+
+impl crate::HashEngine for HashEngine {
+    type MidState = [u8; 64];
+
+    #[cfg(not(fuzzing))]
+    fn midstate(&self) -> [u8; 64] {
+        let mut ret = [0; 64];
+        for (val, ret_bytes) in self.0.h.iter().zip(ret.chunks_exact_mut(8)) {
+            ret_bytes.copy_from_slice(&val.to_be_bytes());
+        }
+        ret
+    }
+
+    #[cfg(fuzzing)]
+    fn midstate(&self) -> [u8; 64] {
+        let mut ret = [0; 64];
+        ret.copy_from_slice(&self.buffer[..64]);
+        ret
+    }
+
+    const BLOCK_SIZE: usize = 128;
+
+    fn n_bytes_hashed(&self) -> usize {
+        self.0.length
+    }
+
+    #[cfg(not(fuzzing))]
+    fn input(&mut self, mut inp: &[u8]) {
+        while !inp.is_empty() {
+            let buf_idx = self.0.length % <Self as crate::HashEngine>::BLOCK_SIZE;
+            let rem_len = <Self as crate::HashEngine>::BLOCK_SIZE - buf_idx;
+            let write_len = cmp::min(rem_len, inp.len());
+
+            self.0.buffer[buf_idx..buf_idx + write_len]
+                .copy_from_slice(&inp[..write_len]);
+            self.0.length += write_len;
+            if self.0.length % <Self as crate::HashEngine>::BLOCK_SIZE == 0 {
+                self.0.process_block();
+            }
+            inp = &inp[write_len..];
+        }
+    }
+
+    #[cfg(fuzzing)]
+    fn input(&mut self, inp: &[u8]) {
+        for c in inp {
+            self.0.buffer[0] ^= *c;
+        }
+        self.0.length += inp.len();
+    }
```

Comment:
Isn't this the same as `self.0.input(inp)`?

Reply:
Yup. Addressed.